File: version4j/src/main/java/com/only4/algorithm/version4j/leetcode/MinStack.java
package com.only4.algorithm.version4j.leetcode;

import java.util.ArrayDeque;

class MinStack {
    private final ArrayDeque<Node> stack;

    static class Node {
        int val;
        int min;

        Node(int val, int min) {
            this.val = val;
            this.min = min;
        }
    }

    public MinStack() {
        stack = new ArrayDeque<>();
    }

    public void push(int val) {
        int currentMin = stack.isEmpty() ? val : Math.min(val, getMin());
        stack.addLast(new Node(val, currentMin));
    }

    public void pop() {
        stack.removeLast();
    }

    public int top() {
        return stack.peekLast().val;
    }

    public int getMin() {
        return stack.peekLast().min;
    }
}

File: version4k/src/main/kotlin/com/only4/algorithm/version4k/leetcode/MinStack.kt
package com.only4.algorithm.version4k.leetcode

class MinStack {
    private val stack = ArrayDeque<Pair<Int, Int>>()

    fun push(value: Int) {
        val currentMin = if (stack.isEmpty()) value else minOf(getMin(), value)
        stack.addLast(Pair(value, currentMin))
    }

    fun pop() {
        stack.removeLast()
    }

    fun top(): Int = stack.last().first

    fun getMin(): Int = stack.last().second
}
